### PR TITLE
Make changes to support building Mono for IL2CPP

### PR DESCRIPTION
**Pull request by joshpeterson**
These changes allow us to use libmonoruntime with IL2CPP.

* The changes here are mostly additive. Any change in Mono either have no impact or are #ifdefed out.
* The submodule for external/buildscipts points to the monobuildtools repo at unity-mono-4.8.0-branch-il2cpp. Once this PR lands, I'll merge that branch into unity-mono-4.8.0-branch and update he submodule in this repo.

### CODE REVIEW

**Comment by joshpeterson** [comment]
Once this PR lands, and we get the binaries into il2cpp-dependencies, we will PR the IL2CPP changes to default.

File: mono/metadata/class.h [hunk 1]
```diff
@@ -268,6 +268,12 @@ mono_method_can_access_field (MonoMethod *method, MonoClassField *field);
 MONO_API mono_bool
 mono_method_can_access_method (MonoMethod *method, MonoMethod *called);
 
+MONO_API MonoMethod*
+mono_method_get_method_definition(MonoMethod *method);
+
+MONO_API void
+mono_class_set_allow_gc_aware_layout(mono_bool allow);
```
Comment (joncham) [inline mono/metadata/class.h]: Is it worth exposing this, or should we just set it based on IL2CPP being used?

File: mono/metadata/threadpool-ms.c [hunk 2]
```diff
@@ -749,7 +749,9 @@ worker_try_create (void)
 	current_ticks = mono_100ns_ticks ();
 	now = current_ticks / (10 * 1000 * 1000);
 	if (0 == current_ticks) {
+#ifndef IL2CPP_ON_MONO
```
Comment (joncham) [inline mono/metadata/threadpool-ms.c]: Why does this not work with IL2CPP?

File: msvc/mono.def [hunk 3]
```diff
@@ -966,3 +966,5 @@ unity_mono_reflection_method_get_method
 mono_unity_g_free
 mono_set_find_plugin_callback
 mono_unity_set_vprintf_func
+mono_type_get_name_full
```
Comment (joncham) [inline msvc/mono.def]: Why are these needed? Anything we add here we should add to the sgen .def file as well.

File: mono/metadata/class.h (hunk 1 above)
Comment (joshpeterson) [inline mono/metadata/class.h]: I don't see a reason not to expose the both. `mono_method_get_method_definition` is generally useful, and `mono_class_set_allow_gc_aware_layout` is just a switch.

File: mono/metadata/threadpool-ms.c (hunk 2 above)
Comment (joshpeterson) [inline mono/metadata/threadpool-ms.c]: This warning randomly shows up in the output, and that causes our tests to fail.

File: msvc/mono.def (hunk 3 above)
Comment (joshpeterson) [inline msvc/mono.def]: I don't recall why these were added. @brianradunity do you know? If not, we can try to revert the changes in this file and see what happens.

Comment (brianradunity) [inline msvc/mono.def]: Not sure.  I don't recall touching the .def file.

Comment (joshpeterson) [inline msvc/mono.def]: Ok, I'll try without this change next week.

Comment (joshpeterson) [inline msvc/mono.def]: I'll remove this.

File: mono/metadata/class.h (hunk 1 above)
Comment (joncham) [inline mono/metadata/class.h]: I think it's safe to remove MONO_API here if these aren't in the def file and things work.

Comment (joshpeterson) [inline mono/metadata/class.h]: I'll remove it.
